Overlay inserting docset into array item results in panic

**What steps did you take:**
Applied an overlay over a document.

https://carvel.dev/ytt/#gist:https://gist.github.com/pivotaljohn/6ed947ba79d31449dcb9b61921566429

```yaml
#! base.yml
---
map1:
  map2:
    map3:
    - ""
```

```yaml
#! overlay.yml
#@ load("@ytt:overlay", "overlay")

#@ def make_docset():
---
something: 1
#@ end

#@overlay/match by=overlay.all
---
map1:
  map2:
    map3:
    #@overlay/append
    - #@ make_docset()
```


**What happened:**

```
$ ytt -f base.yml -f overlay.yml
panic: Unexpected docset value within document
goroutine 1 [running]:
github.com/k14s/ytt/pkg/yamlmeta.convertToGo(0x30bd220, 0xc0000327e0, 0x0, 0x0)
        /root/go/pkg/mod/github.com/k14s/ytt@v0.30.0/pkg/yamlmeta/convert.go:52 +0x734
github.com/k14s/ytt/pkg/yamlmeta.convertToGo(0x3017740, 0xc001068730, 0xc000a22690, 0x0)
        /root/go/pkg/mod/github.com/k14s/ytt@v0.30.0/pkg/yamlmeta/convert.go:71 +0x47e
github.com/k14s/ytt/pkg/yamlmeta.convertToGo(0x30179c0, 0xc00102a960, 0xc000a22400, 0x0)
        /root/go/pkg/mod/github.com/k14s/ytt@v0.30.0/pkg/yamlmeta/convert.go:64 +0x5fd
github.com/k14s/ytt/pkg/yamlmeta.convertToGo(0x30179c0, 0xc001025770, 0xc000a22380, 0x0)
        /root/go/pkg/mod/github.com/k14s/ytt@v0.30.0/pkg/yamlmeta/convert.go:64 +0x5fd
github.com/k14s/ytt/pkg/yamlmeta.convertToGo(0x30179c0, 0xc000ff8050, 0x4f96d8, 0xc0011b6c8e)
        /root/go/pkg/mod/github.com/k14s/ytt@v0.30.0/pkg/yamlmeta/convert.go:64 +0x5fd
github.com/k14s/ytt/pkg/yamlmeta.(*Document).AsYAMLBytes(0xc000fd2410, 0xc00050d6f8, 0x4, 0x4, 0x4, 0x0)
        /root/go/pkg/mod/github.com/k14s/ytt@v0.30.0/pkg/yamlmeta/document.go:22 +0x3c
github.com/k14s/ytt/pkg/yamlmeta.(*YAMLPrinter).Print(0xc00113b620, 0xc000fd2410, 0x0, 0x0)
        /root/go/pkg/mod/github.com/k14s/ytt@v0.30.0/pkg/yamlmeta/printers.go:33 +0x42
github.com/k14s/ytt/pkg/yamlmeta.(*DocumentSet).AsBytesWithPrinter(0xc0003e58f0, 0x3875c40, 0x7, 0x8, 0xc0010ae0c0, 0x7, 0x8)
        /root/go/pkg/mod/github.com/k14s/ytt@v0.30.0/pkg/yamlmeta/document_set.go:55 +0xec
github.com/k14s/ytt/pkg/yamlmeta.(*DocumentSet).AsBytes(...)
        /root/go/pkg/mod/github.com/k14s/ytt@v0.30.0/pkg/yamlmeta/document_set.go:40
github.com/k14s/ytt/pkg/workspace.(*LibraryLoader).Eval(0xc0014bacf0, 0xc000e7bbc0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
        /root/go/pkg/mod/github.com/k14s/ytt@v0.30.0/pkg/workspace/library_loader.go:111 +0x33c
```

translated:
- "the document contains a map
- ... that has a map item containing a map
- ... that has a map item containing a map
- ... that has a map item containing an array
- ... that has an array item containing a document set."

... with the help of these code snippets:

https://github.com/vmware-tanzu/carvel-ytt/blob/v0.30.0/pkg/yamlmeta/convert.go#L57-L66
https://github.com/vmware-tanzu/carvel-ytt/blob/v0.30.0/pkg/yamlmeta/convert.go#L68-L73
https://github.com/vmware-tanzu/carvel-ytt/blob/v0.30.0/pkg/yamlmeta/convert.go#L51-L52


**What did you expect:**
A user-friendly error message _when the overlay was applying_ that a document set is the inappropriate value to insert that the specified location.

Using current error message formatting, something like this:
```
Overlaying (in following order: overlay.yml):
  Document on line overlay.yml:9:
    Map item (key 'map1') on line overlay.yml:10:
      Map item (key 'map2') on line overlay.yml:11:
        Map item (key 'map3') on line overlay.yml:12:
          Array item (index 0) on line overlay.yml:14:
            Expected docset, but was *yamlmeta.Array
```

**Anything else you would like to add:**

Was reported by @rosenhouse.

**Environment:**

- ytt version (use `ytt --version`): v0.30.0
- OS (e.g. from `/etc/os-release`):

--- a/pkg/cmd/template/cmd_test.go
+++ b/pkg/cmd/template/cmd_test.go
@@ -826,3 +826,84 @@ func TestLoadYTTModuleFailEarly(t *testing.T) {
 		t.Fatalf("Expected RunWithFiles to fail with error, but was '%s'", out.Err.Error())
 	}
 }
+
+func TestBuildingInvalidYAMLFailsFast(t *testing.T) {
+	t.Run("setting document value to DocumentSet", func(t *testing.T) {
+		tpl := []byte(`
+#@ def/end docset():
+---
+foo: true
+
+--- #@ docset()
+`)
+
+		filesToProcess := files.NewSortedFiles([]*files.File{
+			files.MustNewFileFromSource(files.NewBytesSource("template.yml", tpl)),
+		})
+
+		ui := ui.NewTTY(false)
+		opts := cmdtpl.NewOptions()
+
+		out := opts.RunWithFiles(cmdtpl.Input{Files: filesToProcess}, ui)
+		if out.Err == nil {
+			t.Fatalf("Expected RunWithFiles to fail")
+		}
+
+		if !strings.Contains(out.Err.Error(), "documents can only contain arrays, maps, or scalars") {
+			t.Fatalf("Expected RunWithFiles to fail with error, but was '%s'", out.Err.Error())
+		}
+	})
+	t.Run("setting map item to DocumentSet", func(t *testing.T) {
+		tpl := []byte(`
+#@ def/end docset():
+---
+foo: true
+
+---
+bar: #@ docset()
+`)
+
+		filesToProcess := files.NewSortedFiles([]*files.File{
+			files.MustNewFileFromSource(files.NewBytesSource("template.yml", tpl)),
+		})
+
+		ui := ui.NewTTY(false)
+		opts := cmdtpl.NewOptions()
+
+		out := opts.RunWithFiles(cmdtpl.Input{Files: filesToProcess}, ui)
+		if out.Err == nil {
+			t.Fatalf("Expected RunWithFiles to fail")
+		}
+
+		if !strings.Contains(out.Err.Error(), "mapitems can only contain arrays, maps, or scalars") {
+			t.Fatalf("Expected RunWithFiles to fail with error, but was '%s'", out.Err.Error())
+		}
+	})
+	t.Run("setting array item to DocumentSet", func(t *testing.T) {
+		tpl := []byte(`
+#@ def/end docset():
+---
+foo: true
+
+---
+bar:
+- #@ docset()
+`)
+
+		filesToProcess := files.NewSortedFiles([]*files.File{
+			files.MustNewFileFromSource(files.NewBytesSource("template.yml", tpl)),
+		})
+
+		ui := ui.NewTTY(false)
+		opts := cmdtpl.NewOptions()
+
+		out := opts.RunWithFiles(cmdtpl.Input{Files: filesToProcess}, ui)
+		if out.Err == nil {
+			t.Fatalf("Expected RunWithFiles to fail")
+		}
+
+		if !strings.Contains(out.Err.Error(), "arrayitems can only contain maps, arrays, or scalars") {
+			t.Fatalf("Expected RunWithFiles to fail with error, but was '%s'", out.Err.Error())
+		}
+	})
+}

--- a/pkg/yamlmeta/node.go
+++ b/pkg/yamlmeta/node.go
@@ -8,6 +8,7 @@ import (
 	"fmt"
 
 	"github.com/k14s/ytt/pkg/filepos"
+	"github.com/k14s/ytt/pkg/orderedmap"
 	"github.com/k14s/ytt/pkg/yamlmeta/internal/yaml.v2"
 )
 
@@ -48,8 +49,8 @@ func (ds *DocumentSet) SetValue(val interface{}) error {
 }
 
 func (d *Document) SetValue(val interface{}) error {
-	d.Value = val
-	return nil
+	d.ResetValue()
+	return d.AddValue(val)
 }
 
 func (m *Map) SetValue(val interface{}) error {
@@ -57,11 +58,8 @@ func (m *Map) SetValue(val interface{}) error {
 }
 
 func (mi *MapItem) SetValue(val interface{}) error {
-	if isMapOrArrayItem(val) {
-		return fmt.Errorf("cannot set map-or-array-item value (%T) into mapitem", val)
-	}
-	mi.Value = val
-	return nil
+	mi.ResetValue()
+	return mi.AddValue(val)
 }
 
 func (a *Array) SetValue(val interface{}) error {
@@ -69,11 +67,24 @@ func (a *Array) SetValue(val interface{}) error {
 }
 
 func (ai *ArrayItem) SetValue(val interface{}) error {
-	if isMapOrArrayItem(val) {
-		return fmt.Errorf("cannot set map-or-array-item value (%T) into arrayitem", val)
+	ai.ResetValue()
+	return ai.AddValue(val)
+}
+
+func isValidValue(val interface{}) bool {
+	switch val.(type) {
+	case *Map, *orderedmap.Map,
+		*Array, []interface{},
+		int, int8, int16, int32, int64,
+		uint, uint8, uint16, uint32, uint64,
+		float32, float64,
+		bool,
+		string,
+		nil:
+		return true
+	default:
+		return false
 	}
-	ai.Value = val
-	return nil
 }
 
 func (ds *DocumentSet) ResetValue() { ds.Items = nil }
@@ -92,8 +103,8 @@ func (ds *DocumentSet) AddValue(val interface{}) error {
 }
 
 func (d *Document) AddValue(val interface{}) error {
-	if isMapOrArrayItem(val) {
-		return fmt.Errorf("cannot add map-or-array-item value (%T) into document", val)
+	if !isValidValue(val) {
+		return fmt.Errorf("documents can only contain arrays, maps, or scalars; this is a %T", val)
 	}
 	d.Value = val
 	return nil
@@ -108,8 +119,8 @@ func (m *Map) AddValue(val interface{}) error {
 }
 
 func (mi *MapItem) AddValue(val interface{}) error {
-	if isMapOrArrayItem(val) {
-		return fmt.Errorf("cannot add map-or-array-item value (%T) into mapitem", val)
+	if !isValidValue(val) {
+		return fmt.Errorf("mapitems can only contain arrays, maps, or scalars; this is a %T", val)
 	}
 	mi.Value = val
 	return nil
@@ -124,20 +135,11 @@ func (a *Array) AddValue(val interface{}) error {
 }
 
 func (ai *ArrayItem) AddValue(val interface{}) error {
-	if isMapOrArrayItem(val) {
-		return fmt.Errorf("cannot add map-or-array-item value (%T) into arrayitem", val)
+	if !isValidValue(val) {
+		return fmt.Errorf("arrayitems can only contain maps, arrays, or scalars; this is a %T", val)
 	}
 	ai.Value = val
 	return nil
-}
-
-func isMapOrArrayItem(val interface{}) bool {
-	switch val.(type) {
-	case *MapItem, *ArrayItem:
-		return true
-	default:
-		return false
-	}
 }
 
 func (ds *DocumentSet) GetValues() []interface{} {

--- a/pkg/yttlibrary/overlay/array.go
+++ b/pkg/yttlibrary/overlay/array.go
@@ -110,7 +110,10 @@ func (o Op) replaceArrayItem(
 		}
 
 		leftArray.Items[leftIdx] = newItem.DeepCopy()
-		leftArray.Items[leftIdx].SetValue(newVal)
+		err = leftArray.Items[leftIdx].SetValue(newVal)
+		if err != nil {
+			return err
+		}
 	}
 
 	return nil

--- a/pkg/yttlibrary/overlay/document.go
+++ b/pkg/yttlibrary/overlay/document.go
@@ -108,7 +108,10 @@ func (o Op) replaceDocument(
 		}
 
 		leftDocSets[leftIdx[0]].Items[leftIdx[1]] = newDoc.DeepCopy()
-		leftDocSets[leftIdx[0]].Items[leftIdx[1]].SetValue(newVal)
+		err = leftDocSets[leftIdx[0]].Items[leftIdx[1]].SetValue(newVal)
+		if err != nil {
+			return err
+		}
 	}
 
 	return nil

--- a/pkg/yttlibrary/overlay/map.go
+++ b/pkg/yttlibrary/overlay/map.go
@@ -109,7 +109,10 @@ func (o Op) replaceMapItem(leftMap *yamlmeta.Map, newItem *yamlmeta.MapItem,
 		}
 
 		leftMap.Items[leftIdx] = newItem.DeepCopy()
-		leftMap.Items[leftIdx].SetValue(newVal)
+		err = leftMap.Items[leftIdx].SetValue(newVal)
+		if err != nil {
+			return err
+		}
 	}
 
 	return nil

